aptos-move/aptos-gas-algebra/src/abstract_algebra.rs

```diff
@@ -0,0 +1,273 @@
+// Copyright © Aptos Foundation
+// SPDX-License-Identifier: Apache-2.0
+
+use either::Either;
+use move_core_types::gas_algebra::{GasQuantity, UnitDiv};
+use std::{
+    marker::PhantomData,
+    ops::{Add, Mul},
+};
+
+/***************************************************************************************************
+ * Gas Expression & Visitor
+ *
+ **************************************************************************************************/
+/// Trait representing an abstract expression that can be used to calculate some gas amount.
+///
+/// It carries a type parameter `E`, indicating an environment in which the expression can be
+/// evaluated/materialized.
+pub trait GasExpression<E> {
+    type Unit;
+
+    /// Evaluates the expression within the given environment to a concrete number.
+    fn evaluate(&self, feature_version: u64, env: &E) -> GasQuantity<Self::Unit>;
+
+    /// Traverse the expression in post-order using the given visitor.
+    /// See [`GasExpressionVisitor`] for details.
+    fn visit(&self, visitor: &mut impl GasExpressionVisitor);
+
+    /// Performs a division on the unit of the expression.
+    ///
+    /// This is sometimes required if you want to multiply an amount by a certain count.
+    fn per<U>(self) -> GasPerUnit<Self, U>
+    where
+        Self: Sized,
+    {
+        GasPerUnit {
+            inner: self,
+            phantom: PhantomData,
+        }
+    }
+}
+
+/// An interface for performing post-order traversal of the tree structure of a gas expression.
+///
+/// Alternatively, one could think that the callbacks are invoked following the Reverse Polish
+/// notation of the expression.
+///
+/// Here are a few examples:
+/// - `1 + 2`
+///   - `quantity(1)`
+///   - `quantity(2)`
+///   - `add()`
+/// - `A + B * 50`
+///   - `gas_param<A>()`
+///   - `gas_param<B>()`
+///   - `quantity(50)`
+///   - `mul()`
+///   - `add()`
+pub trait GasExpressionVisitor {
+    fn add(&mut self);
+
+    fn mul(&mut self);
+
+    fn gas_param<P>(&mut self);
```

Comment:
is this used anywhere? if so we should probably unify the name with `env` (from what i understand is the params)

Reply:
1. Unused right now, but required for the gas calibrator & future analysis tools.
2. `P` is not the same as `env`. It is instantiated with types representing each individual gas parameters. When you add a gas parameter to `aptos-gas-schedule`, my macro generates a new type for it (e.g. LD_CONST_BASE), which you can use in gas expressions.

Reply:
so it's individual gas param while env is the collection of all? it's not very clear to me what this function means for a visitor though

Reply:
Ah, take a look at the doc comments I wrote for this function above. The visitor callbacks are invoked in an order following the Reverse Polish representation of the expression.

Reply:
yes I understand the order of invoke, but I don't know what it means to call `gas_param`

Reply:
If you see the arithmetic expression as an AST, then `gas_param<B>` is called when we are visiting a node corresponding to gas parameter B.
```
A + B * 5
    ^ It's this B
```

Reply:
Unclear what the `P` is needed for, specifically as it is nowhere used in this trait.

Reply:
Again, all of the methods in the visitor trait are callbacks.

You as the client implement a visitor, and I as the VM will call your methods in a particular sequence, traversing a gas expression.

Here `P` is just a function parameter that tells you (the client) which gas parameter it is currently visiting. Conceptually it is no different from
```
fn gas_param(name: &str);
```
It's just that in this static representation, we're using rust types, rather than strings to name gas parameters.